src/TesApi.Web/Scheduler.cs

```diff
@@ -204,36 +214,117 @@ private async Task ProcessTesTasksContinuouslyAsync(Expression<Func<TesTask, boo
+                    }
+
+                    // Set class-level object so that it can be used by the existing task processor thread
+                    this.completedTesTaskIds = completedTesTaskIds;
+
+                    if (completedTesTaskIds.Count > 0)
+                    {
+                        logger.LogInformation($"{completedTesTaskIds.Count} TES task completions downloaded in {sw.Elapsed.TotalSeconds:n1}s; {sw.Elapsed.TotalSeconds / completedTesTaskIds.Count:n1} seconds per task completion");
+                    }
+
+                    await Task.Delay(runInterval, stoppingToken);                    
                 }
-                catch (RepositoryCollisionException exc)
+                catch (OperationCanceledException) when (stoppingToken.IsCancellationRequested)
                 {
-                    logger.LogError(exc, $"RepositoryCollisionException in OrchestrateTesTasksOnBatch");
+                    logger.LogInformation($"ProcessTesTaskCompletionsContinuouslyAsync tasks cancelled.");
+                    break;
                 }
                 catch (Exception exc)
                 {
-                    logger.LogError(exc, "Updating TES Task '{TesTask}' threw {ExceptionType}: '{ExceptionMessage}'. Stack trace: {ExceptionStackTrace}", tesTask.Id, exc.GetType().FullName, exc.Message, exc.StackTrace);
+                    logger.LogError(exc, exc.Message);
+                    await Task.Delay(runInterval, stoppingToken);
                 }
+            }
+        }
 
-                if (!string.IsNullOrWhiteSpace(tesTask.PoolId) && (TesState.QUEUEDEnum == tesTask.State || TesState.RUNNINGEnum == tesTask.State))
+        private async Task ProcessTesTasks(List<TesTask> tesTasks, bool areExistingTesTasks, CancellationToken stoppingToken)
+        {
+            var pools = new HashSet<string>();
+            int completedTesTaskIdsCount = completedTesTaskIds.Count;
+
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                // Prioritize processing completed tasks, then terminal state tasks, then randomize order within each state to prevent head-of-line blocking
+                foreach (var tesTask in tesTasks.OrderByDescending(t => completedTesTaskIds.Contains(t.Id)).ThenBy(t => t.State).ThenBy(t => random.Next()))
                 {
-                    pools.Add(tesTask.PoolId);
+                    if (areExistingTesTasks && processNewTasksLock.CurrentCount == 0)
+                    {
+                        // New tasks found, stop processing existing tasks
+                        return;
+                    }
+
+                    if (completedTesTaskIds.Count > completedTesTaskIdsCount)
+                    {
+                        // New task completions found, stop processing existing tasks
+                        // Force loop to restart so that new task completions are processed
+                        break;
```

Comment:
fyi; this is how new TES task completions force restarting the loop and processing them first